File: typescript/src/integrations/mcp.ts
/**
 * MCP integration: feed Descope-vaulted tokens into an MCP client's OAuth seam.
 *
 * The MCP spec defines a pluggable `OAuthClientProvider` that an MCP *client* uses
 * to obtain the token it sends to a remote MCP *server*. Normally the client runs
 * the whole OAuth dance itself (discovery, DCR, the authorization-code redirect,
 * storage, refresh). In a *brokered* model, Descope already holds and refreshes that
 * token in the vault — so this provider short-circuits the dance: `tokens()` returns
 * the vaulted token and the rest are deliberate no-ops.
 *
 * The object returned is structurally compatible with the `OAuthClientProvider` in
 * `@modelcontextprotocol/sdk` and the `authProvider` accepted by `@ai-sdk/mcp`'s
 * `createMCPClient` — so you can pass it straight in, and this package keeps a zero
 * dependency on the MCP SDK.
 *
 *     import { createMCPClient } from '@ai-sdk/mcp';
 *     import { descopeMcpConnectionAuthProvider } from '@descope/agent-auth';
 *
 *     const mcp = await createMCPClient({
 *       transport: {
 *         type: 'http',
 *         url: 'https://mcp.linear.app',
 *         authProvider: descopeMcpConnectionAuthProvider(client, {
 *           connection: 'linear',
 *           identifier: userId,        // resolved server-side, never from the model
 *         }),
 *       },
 *     });
 *
 * When the user hasn't connected the provider yet, `tokens()` throws
 * `ConnectionAuthorizationRequired` (carrying `connectUrl`) — catch it where you run
 * the agent and redirect the user to consent, exactly as you would for a direct
 * `connections.getToken` call. After they consent, the next call just works.
 *
 * Two flavors, matching the two token types:
 *  - `descopeMcpConnectionAuthProvider` — the MCP server is protected by a provider
 *    you've set up as a Descope Connection (the token is that provider's own token).
 *  - `descopeMcpResourceAuthProvider` — the MCP server treats Descope as its OAuth
 *    authorization server (the token is a Descope-minted Resource token).
 */

import type { AgentAuthClient } from '../client';
import type { ApprovalRequest } from '../types';

/** Minimal OAuth token shape the MCP transports read (subset of MCP's `OAuthTokens`). */
export interface McpOAuthTokens {
  access_token: string;
  token_type: string;
  scope?: string;
}

/**
 * The subset of the MCP-spec `OAuthClientProvider` a brokered provider implements.
 * Structurally assignable to the real interface, so it can be passed to `authProvider`
 * without importing `@modelcontextprotocol/sdk`.
 */
export interface McpAuthProvider {
  readonly redirectUrl: string | URL | undefined;
  readonly clientMetadata: { redirect_uris: string[] };
  clientInformation(): undefined;
  tokens(): Promise<McpOAuthTokens>;
  saveTokens(): void;
  redirectToAuthorization(): void;
  saveCodeVerifier(): void;
  codeVerifier(): string;
}

export interface McpConnectionAuthOptions {
  /** The Descope Connection / Outbound App name (e.g. `'linear'`, `'github'`). */
  connection: string;
  /** The user the agent acts for. Resolve this server-side — never from model input. */
  identifier: string;
  /** Override the Connection's default scopes (fully replaces them when set). */
  scopes?: string[];
  /** Select a user's per-tenant token. */
  tenantId?: string;
  /** Act as a specific user on a shared client (the user's Descope access token). */
  actAsUserToken?: string;
  /** Require a fresh CIBA approval (needs an `approval` provider on the client). */
  requireApproval?: ApprovalRequest;
  /** Reported as the provider's `redirectUrl`; the brokered flow doesn't use it. */
  redirectUrl?: string;
}

export interface McpResourceAuthOptions {
  /** RFC 8707 resource indicator for the API the MCP server represents. */
  resource: string;
  scopes?: string[];
  /** Token-exchange `audience` claim, when the resource requires it. */
  audience?: string[];
  actAsUserToken?: string;
  /** Require a fresh CIBA approval (needs an `approval` provider on the client). */
  requireApproval?: ApprovalRequest;
}

const brokeredProvider = (
  fetchToken: () => Promise<{ accessToken: string; tokenType: string; scopes: string[] }>,
  redirectUrl?: string,
): McpAuthProvider => ({
  // Consent is surfaced via the thrown ConnectionAuthorizationRequired, so the SDK's
  // own redirect machinery is never driven — these stay no-ops by design.
  get redirectUrl() {
    return redirectUrl;
  },
  get clientMetadata() {
    return { redirect_uris: redirectUrl ? [redirectUrl] : [] };
  },
  clientInformation() {
    return undefined;
  },
  async tokens() {
    const t = await fetchToken();
    return {
      access_token: t.accessToken,
      token_type: t.tokenType || 'bearer',
      scope: t.scopes.length ? t.scopes.join(' ') : undefined,
    };
  },
  saveTokens() {
    /* Descope owns storage + refresh */
  },
  redirectToAuthorization() {
    /* brokered: consent happens at the Descope connect URL */
  },
  saveCodeVerifier() {
    /* PKCE handled inside Descope's brokered flow */
  },
  codeVerifier() {
    return '';
  },
});

/**
 * Build an MCP `authProvider` that injects a Descope **Connection** token (the
 * provider's own OAuth token) for the given user.
 */
export function descopeMcpConnectionAuthProvider(
  client: AgentAuthClient,
  options: McpConnectionAuthOptions,
): McpAuthProvider {
  return brokeredProvider(
    () =>
      client.connections.getToken({
        connection: options.connection,
        identifier: options.identifier,
        scopes: options.scopes,
        tenantId: options.tenantId,
        actAsUserToken: options.actAsUserToken,
        requireApproval: options.requireApproval,
      }),
    options.redirectUrl,
  );
}

/**
 * Build an MCP `authProvider` that injects a Descope-minted **Resource** token (for
 * an MCP server that uses Descope as its OAuth authorization server).
 */
export function descopeMcpResourceAuthProvider(
  client: AgentAuthClient,
  options: McpResourceAuthOptions,
): McpAuthProvider {
  return brokeredProvider(() =>
    client.resources.getToken({
      resource: options.resource,
      scopes: options.scopes,
      audience: options.audience,
      actAsUserToken: options.actAsUserToken,
      requireApproval: options.requireApproval,
    }),
  );
}

File: python/descope_agent_auth/integrations/mcp.py
"""MCP integration: feed Descope-vaulted tokens into an MCP client's HTTP auth.

The MCP Python client connects to a remote MCP server over an ``httpx``-based
transport. Rather than have the client run the whole OAuth dance itself (discovery,
DCR, the authorization-code redirect, storage, refresh), a *brokered* setup lets
Descope hold and refresh that token in the vault -- and this module hands the
transport an ``httpx.Auth`` that injects the vaulted token as a ``Bearer`` header.

An ``httpx.Auth`` is the portable integration point: the MCP transport accepts one
directly (``auth=``) on releases that expose it, and on newer ones you set it on the
``httpx.AsyncClient`` you pass in -- either way the same object works::

    from mcp.client.streamable_http import streamablehttp_client  # released API
    from descope_agent_auth import AsyncAgentAuthClient
    from descope_agent_auth.integrations.mcp import connection_auth

    auth = connection_auth(client, connection="linear", identifier=user_id)
    async with streamablehttp_client("https://mcp.linear.app", auth=auth) as (r, w, _):
        ...

When the user hasn't connected the provider yet, the token fetch raises
``ConnectionAuthorizationRequired`` (carrying ``connect_url``); it propagates out of
the request so you can catch it where you drive the agent and send the user to
consent, exactly as with a direct ``connections.get_token`` call.

Two flavors, matching the two token types:
  - :func:`connection_auth` -- the MCP server is protected by a provider you've set
    up as a Descope Connection (the token is that provider's own token).
  - :func:`resource_auth` -- the MCP server treats Descope as its OAuth authorization
    server (the token is a Descope-minted Resource token).

This module depends only on ``httpx`` (already a core dependency); it never imports
the MCP SDK.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, AsyncGenerator, Awaitable, Callable, Generator, List, Optional

import httpx

if TYPE_CHECKING:
    from ..client import AsyncAgentAuthClient

_TokenFetcher = Callable[[], Awaitable[str]]


class _BearerInjectingAuth(httpx.Auth):
    """An ``httpx.Auth`` that fetches a Descope-vaulted token per request and sends
    it as ``Authorization: Bearer ...``. Async-only -- the MCP client is async."""

    requires_request_body = False
    requires_response_body = False

    def __init__(self, fetch: _TokenFetcher) -> None:
        self._fetch = fetch

    def sync_auth_flow(
        self, request: httpx.Request
    ) -> Generator[httpx.Request, httpx.Response, None]:
        raise RuntimeError(
            "Descope MCP auth is async-only; use an async MCP client / httpx.AsyncClient."
        )

    async def async_auth_flow(
        self, request: httpx.Request
    ) -> AsyncGenerator[httpx.Request, httpx.Response]:
        # The token fetch is cached + refreshed inside Descope, so per-request calls
        # are cheap. ConnectionAuthorizationRequired (consent needed) propagates here.
        token = await self._fetch()
        request.headers["Authorization"] = f"Bearer {token}"
        yield request


def connection_auth(
    client: "AsyncAgentAuthClient",
    *,
    connection: str,
    identifier: str,
    scopes: Optional[List[str]] = None,
    tenant_id: Optional[str] = None,
    act_as_user_token: Optional[str] = None,
) -> httpx.Auth:
    """Build an ``httpx.Auth`` that injects a Descope **Connection** token (the
    provider's own OAuth token) for ``identifier`` on each MCP request.

    ``identifier`` is the user the agent acts for -- resolve it server-side, never
    from model input.
    """

    async def fetch() -> str:
        token = await client.connections.get_token(
            connection=connection,
            identifier=identifier,
            scopes=scopes,
            tenant_id=tenant_id,
            act_as_user_token=act_as_user_token,
        )
        return token.access_token

    return _BearerInjectingAuth(fetch)


def resource_auth(
    client: "AsyncAgentAuthClient",
    *,
    resource: str,
    scopes: Optional[List[str]] = None,
    audience: Optional[List[str]] = None,
    act_as_user_token: Optional[str] = None,
) -> httpx.Auth:
    """Build an ``httpx.Auth`` that injects a Descope-minted **Resource** token (for
    an MCP server that uses Descope as its OAuth authorization server)."""

    async def fetch() -> str:
        token = await client.resources.get_token(
            resource=resource,
            scopes=scopes,
            audience=audience,
            act_as_user_token=act_as_user_token,
        )
        return token.access_token

    return _BearerInjectingAuth(fetch)
